google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java

```diff
@@ -0,0 +1,251 @@
+      Map<String, String> attributes,
+      StatsWrapper statsWrapper,
+      @Nullable StatsRecorderWrapper statsRecorderWrapper) {
+    this.operationType = operationType;
+    this.spanName = spanName;
+    if (statsRecorderWrapper != null) {
+      // A workaround for test to pass in a mock StatsRecorderWrapper
+      this.recorder = statsRecorderWrapper;
+    } else {
+      this.recorder = new StatsRecorderWrapper(operationType, spanName, attributes, statsWrapper);
+    }
+  }
+
+  @Override
+  public Scope inScope() {
+    return new Scope() {
+      @Override
+      public void close() {}
+    };
+  }
+
+  @Override
+  public void operationSucceeded() {
+    recordOperationCompletion(null);
+  }
+
+  @Override
+  public void operationCancelled() {
+    recordOperationCompletion(new CancellationException());
+  }
+
+  @Override
+  public void operationFailed(Throwable error) {
+    recordOperationCompletion(error);
+  }
+
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    attemptStarted(null, attemptNumber);
+  }
+
+  @Override
+  public void attemptStarted(Object request, int attemptNumber) {
+    this.attempt = attemptNumber;
+    attemptCount++;
+    attemptTimer = Stopwatch.createStarted();
+    if (request != null) {
+      this.tableId = Util.extractTableId(request);
+    }
+    if (applicationLatencyTimerIsRunning.compareAndSet(true, false)) {
+      totalApplicationLatency.addAndGet(applicationLatencyTimer.elapsed(TimeUnit.MILLISECONDS));
+      applicationLatencyTimer.reset();
+    }
+  }
+
+  @Override
+  public void attemptSucceeded() {
+    recordAttemptCompletion(null);
+  }
+
+  @Override
+  public void attemptCancelled() {
+    recordAttemptCompletion(new CancellationException());
+  }
+
+  @Override
+  public void attemptFailed(Throwable error, Duration delay) {
+    if (applicationLatencyTimerIsRunning.compareAndSet(false, true)) {
+      applicationLatencyTimer.start();
+    }
+    recordAttemptCompletion(error);
+  }
+
+  @Override
+  public void attemptFailedRetriesExhausted(Throwable error) {
+    super.attemptFailedRetriesExhausted(error);
+  }
+
+  @Override
+  public void attemptPermanentFailure(Throwable error) {
+    super.attemptPermanentFailure(error);
+  }
+
+  @Override
+  public void lroStartFailed(Throwable error) {
+    super.lroStartFailed(error);
+  }
+
+  @Override
+  public void lroStartSucceeded() {
+    super.lroStartSucceeded();
+  }
+
+  @Override
+  public void onRequest() {
+    if (applicationLatencyTimerIsRunning.compareAndSet(true, false)) {
+      totalApplicationLatency.addAndGet(applicationLatencyTimer.elapsed(TimeUnit.MILLISECONDS));
+      applicationLatencyTimer.reset();
+    }
+  }
+
+  @Override
+  public void responseReceived() {
+    if (applicationLatencyTimerIsRunning.compareAndSet(false, true)) {
+      applicationLatencyTimer.start();
+    }
+    if (firstResponsePerOpTimer.isRunning()) {
+      firstResponsePerOpTimer.stop();
+    }
+  }
```

Comment:
I think your application latency calculations are a bit off. What happens when the caller disables auto flow control and request(10)?